api/src/main/java/com/epam/pipeline/manager/cluster/performancemonitoring/ResourceMonitoringManager.java

```diff
@@ -434,5 +436,30 @@ public void updateInstanceMap(Map<String, InstanceType> types) {
             instanceTypeMap.clear();
             instanceTypeMap.putAll(types);
         }
+
+        private void processLongPausedRuns() {
+            final LongPausedRunAction action = LongPausedRunAction.valueOf(preferenceManager.getPreference(
+                    SystemPreferences.SYSTEM_LONG_PAUSED_ACTION));
+
+            final List<PipelineRun> pausedRuns = pipelineRunManager
+                    .loadRunsByStatuses(Collections.singletonList(TaskStatus.PAUSED)).stream()
+                    .map(run -> pipelineRunManager.loadPipelineRunWithStatuses(run.getId()))
+                    .collect(Collectors.toList());
+
+            processLongPausedRuns(pausedRuns, action);
+        }
+
+        private void processLongPausedRuns(final List<PipelineRun> pausedRuns, final LongPausedRunAction action) {
+            if (CollectionUtils.isEmpty(pausedRuns)) {
+                return;
+            }
+
+            if (LongPausedRunAction.STOP.equals(action)) {
+                ListUtils.emptyIfNull(notificationManager.notifyLongPausedRunsBeforeStop(pausedRuns))
+                        .forEach(run -> pipelineRunManager.stop(run.getId()));
```

Comment:
Just `stop` is not enough in this case as paused instance will stay in the cloud: I'd suggest to use `NodesManager.terminateNode` method here.